### PR TITLE
WFS performing binary spatial operations in expressions on server

### PR DESCRIPTION
**Pull request by signedav**
QGIS used to be able to perform the expression filters of binary spatial operations on the server when the geometry was provided as WKT or in the GML format. With this change it converts the geometry when being a static variable in the expression to an OGC filter. The WFS URL is then created combining bbox rectangle, layer-filter and the filtering expression.

A feature request computes the expression in advance to find out, if it can be performed on the server. Then it's stored in `QgsBackgroundCachedSharedData` where a new download is triggered.

When a filter requests is not a pure binary spatial operation, it will not be evaluated on server. This could be future-improved using partial evaluation on server and passing the untranslated part back.

**Open things:**
- [ ] On having such aggregate operations in expressions for field calculation or symbolization this implementation leads to a lot of downloads (per each feature, since everyone of them has an expression) what can lead to long waiting time (pseudo freeze). Same Situation on other providers. Do we need to consider it?
- [ ] Do we need to consider that GET Requests are too long https://github.com/qgis/QGIS/issues/32348 ?
  This is not new since we where able to pass WKT geometries.

Sponsored by Kanton Schaffhausen

### CODE REVIEW

File: src/core/qgsogcutils.cpp [hunk 1]
```diff
@@ -2368,6 +2372,13 @@ QDomElement QgsOgcUtilsExprToFilter::expressionFunctionToOgcFilter( const QgsExp
       const QDomNode geomNode = mDoc.importNode( geomDoc.documentElement(), true );
       otherGeomElem = geomNode.toElement();
     }
+    else if ( otherNode->hasCachedStaticValue() && otherNode->cachedStaticValue().canConvert<QgsGeometry>() )
+    {
+      QgsGeometry geom = otherNode->cachedStaticValue().value<QgsGeometry>();
+      otherGeomElem = QgsOgcUtils::geometryToGML( geom, mDoc, mGMLVersion, mSrsName, mInvertAxisOrientation,
```
Comment (rouault) [inline src/core/qgsogcutils.cpp:2378]: otherGeomElem should probably be tested to check it is not the empty QDomElement() returned when the geometry cannot be converted to GML

File: src/providers/wfs/qgswfsshareddata.cpp [hunk 2]
```diff
@@ -93,6 +93,43 @@ QString QgsWFSSharedData::srsName() const
...
+    honourAxisOrientation = !mURI.ignoreAxisOrientation();
+    gmlVersion = QgsOgcUtils::GML_3_1_0;
+    filterVersion = QgsOgcUtils::FILTER_OGC_1_1;
+  }
+  else
+  {
+    honourAxisOrientation = !mURI.ignoreAxisOrientation();
+    gmlVersion = QgsOgcUtils::GML_3_2_1;
+    filterVersion = QgsOgcUtils::FILTER_FES_2_0;
+  }
```
Comment (rouault) [inline src/providers/wfs/qgswfsshareddata.cpp:117]: those lines, copied&pasted from QgsWFSSharedData::computeFilter(), could be moved in a small helper function

File: src/providers/wfs/qgswfsfeatureiterator.cpp [hunk 3]
```diff
@@ -327,6 +369,7 @@ QUrl QgsWFSFeatureDownloaderImpl::buildURL( qint64 startIndex, long long maxFeat
 
   getFeatureUrl.setQuery( query );
   QgsDebugMsgLevel( QStringLiteral( "WFS GetFeature URL: %1" ).arg( getFeatureUrl.toDisplayString( ) ), 2 );
+  qDebug() << "\n\n" << "WFS GetFeature URL: " << getFeatureUrl.toDisplayString() << "\n\n";
```
Comment (rouault) [inline src/providers/wfs/qgswfsfeatureiterator.cpp:372]: This should be removed as redundant with the above QgsDebugMsgLevel()

**Review by rouault — CHANGES_REQUESTED** [review]

**Comment by rouault** [comment]
fix for new failed test in https://github.com/signedav/QGIS/pull/18

**Review by rouault — APPROVED** [review]

**Comment by signedav** [comment]
Thanks @rouault and @nyalldawson

**Comment by paul-dorsetcouncil** [comment]
@signedav do you know if anyone ever looked into your comment about GET requests being too long? 

We have been finding that a WFS layer is difficult or even impossible to filter for more than about 3 or 4 features because the filter string is making the URL too long. 

I'd like to find out if others are wanting the change and thought you might know if there is much interest?

**Comment by kannes** [comment]
@paul-dorsetcouncil please file that as a bug or feature request. Down here in a merged pull request no one would see it unless by accident (hi!).

**Comment by signedav** [comment]
Thanks @kannes 

@paul-dorsetcouncil I guess you can put your inputs here https://github.com/qgis/QGIS/issues/32348 And btw I haven't worked on this issue since.

**Comment by kannes** [comment]
Thanks right back at you, @signedav

It looks like @paul-dorsetcouncil already posted there, so this comment thread here can happily go back to sleep. :)

**Pull request merged**